aicssegmentation/structure_wrapper/WorkflowEngine.py

```diff
@@ -0,0 +1,139 @@
+        self.steps: list = self.__get_steps()        # List of WorkflowSteps for this workflow
+        self.currentStep: int = 0                 # Next step to execute
+        self.starting_image: np.ndarray = image          # Initial image
+
+    def __get_steps(self) -> list:
+        """
+        Get a list of WorkflowStep objects to perform on the starting image.
+
+        Params:
+            none
+
+        Returns:
+            (list(WorkflowStep)): List of workflow step objects
+       """
+        # TODO: in order for parent fucntionality to work correctly, we should sort these in the list by parent index
+        return parse_config_to_objects(load_workflow_config(self.workflow_name))
+
+    def get_next_step(self) -> WorkflowStep:
+        """
+        Get the next step to be performed
+
+        Params:
+            none
+
+        Returns:
+            (WorkflowStep): next WorkflowStep object to perform on image
+       """
+        return self.steps[self.currentStep]
+
+    def execute_next(self) -> np.ndarray:
+        """
+        Execute the next workflow step.
+
+        Params:
+            none
+
+        Returns:
+            result (np.ndarray): resultant image from running the next workflow step
+       """
+        print("running step")
+        print(self.currentStep)
+        # Pick which image to perform the workflow step on
+        image: np.ndarray = None
+        if self.currentStep == 0:
+            # First image, so use the starting image for the next workflow step
+            image = self.starting_image
+        elif self.isDone():
+            # No more workflow steps to perform
+            # TODO: what to do if done with workflow but execute_next is prompted?
+            # printing message for now
+            print("No steps left to run")
+        else:
+            # First step has been run, so run next workflow step with the result of its parent
+            image = self.get_result(self.get_next_step().parent)
+
+        result: np.ndarray = self.get_next_step().execute(image)
+
+        # Only increment after running step
+        self.currentStep = self.currentStep + 1
+        return result
+
+
+    def get_result(self, step_index: int) -> np.ndarray:
+        """
+        Get the result image for a workflow step.
+        You must call execute() on the workflow step in order to produce a result first before calling this function.
+
+        Params:
+            step_index (int): index of the WorkflowStep in the workflowengine to get the result image of.
+
+        Returns:
+            self.image (np.ndarray): Result of performing workflow step on the given image
+                                     None if step has not been executed yet.
+       """
+        if step_index > self.currentStep:
+            return None # returns None if the WorkflowStep has not been executed.
+        else:
+            return self.steps[step_index].result
+
+    def get_most_recent_result(self) -> np.ndarray:
```

Comment:
This is not an issue. But, more like a question. 

If the workflow is nicely sequential, then the concept "next" and "get_most_recent_result" is straightforward. However, for more complicated workflow, I am not sure how to communicate the meaning of "next" and "get_most_recent_result". For example, https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/conf_lmnb1_interphase.json

In this example, the output of step 2 goes to step 3, 4, 5, then the output of step 3, 4, 5 are merged in step 6. If you are at step 6, what would be a reasonable "get_most_recent_result"? Currently, it returns step 5 output, which is totally fine in certain context. But, i am not sure this is the best option. I would love to hear more thoughts. 

Also, maybe @thao-do has some thoughts from UI/UX perspective?

Reply:
This is also a thought I had while creating this part of the class. You're right that currently if we call `get_most_recent_result()`, it will just grab the image that has most recently been executed. 

To allow for more functionality in the future i also created a `get_result(int: step_index)` function. Given the index of a step in the workflow, we can also grab results that way.

I realize that we may not have perfectly sequential workflows, but if we sort the `WorkflowSteps` list in the `WorkflowEngine` by parent id, and execute each step sequentially, we can still execute complicated workflows without any problems. For the example you provided above, if we have the result of 3, 4, and 5, then call step 6, the results from the previous steps are available to merge in step 6.

I would love to hear @thao-do 's thoughts from the UI perspective as well!